File: banda/IvanRC-banda-Java/src/main/java/modelsBandaMusicos/Banda.java
package modelsBandaMusicos;

import factoryMusicosBanda.FactoryMusicos;

public class Banda {

    int numerosDeMusicos;

    public Banda(int numeroDeMusicos){
        this.numerosDeMusicos = numeroDeMusicos;
    }
    public Musico[] arrayMusicos = null;

    /**
     * función que sirve para rellenar el array de musicos, con musicos aleatorios
     */
    public void inicializarArray(){
        arrayMusicos = new Musico[numerosDeMusicos];
        for(int i=0;i< arrayMusicos.length;i++){
            arrayMusicos[i] = FactoryMusicos.getInstance().createMusicoRandom();
        }
    }

    /**
     * función que sirve para contar el número de trompetistas en el array
     * @return el numero de trompetistas en el array
     */
    public int contarTrompetistas(){
        int contador = 0;
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof Trompetista) contador++;
        }
        return contador;
    }

    /**
     * función que sirve para contar el número de cantantes en el array
     * @return el numero de cantantes en el array
     */
    public int contarCantantes(){
        int contador = 0;
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof Cantante) contador++;
        }
        return contador;
    }

    /**
     * función que sirve para contar el número de guitarristas en el array
     * @return el numero de guitarristas en el array
     */
    public int contarGuitarristas(){
        int contador = 0;
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof Guitarrista) contador++;
        }
        return contador;
    }

    /**
     * función que sirve para contar el número de bajista en el array
     * @return el numero de bajista en el array
     */
    public int contarBajistas(){
        int contador = 0;
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof Bajista) contador++;
        }
        return contador;
    }

    /**
     * función que sirve para contar el número de teclistas en el array
     * @return el numero de teclistas en el array
     */
    public int contarTeclistas(){
        int contador = 0;
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof Teclista) contador++;
        }
        return contador;
    }

    /**
     * función que sirve para contar el número de percusionista en el array
     * @return el numero de percusionista en el array
     */
    public int contarPercusionistas(){
        int contador = 0;
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof Percusionista) contador++;
        }
        return contador;
    }

    /**
     * función que sirve para contar el número de cantantesGuitarristas en el array
     * @return el numero de cantantesGuitarristas en el array
     */
    public int contarCantantesGuitarristas(){
        int contador = 0;
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof CantanteGuitarrista) contador++;
        }
        return contador;
    }

    /**
     * función que sirve para contar el número de multinstrumentalista en el array
     * @return el numero de multinstrumentalista en el array
     */
    public int contarMultinstrumentales(){
        int contador = 0;
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof Multinstrumental) contador++;
        }
        return contador;
    }

    /**
     * función que permite calcular el salario total
     * @return el salario total
     */
    public int hallarSalarioTotal(){
        int resultado = 0;
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof Trompetista){
                resultado += ((Trompetista) arrayMusicos[i]).recalcularSalario(arrayMusicos[i].salario);
            }else{
                if(arrayMusicos[i] instanceof Cantante){
                    resultado += ((Cantante) arrayMusicos[i]).recalcularSalario(arrayMusicos[i].salario);
                }else{
                    if(arrayMusicos[i] instanceof Guitarrista){
                        resultado += ((Guitarrista) arrayMusicos[i]).recalcularSalario(arrayMusicos[i].salario);
                    }else{
                        if(arrayMusicos[i] instanceof Bajista){
                            resultado += ((Bajista) arrayMusicos[i]).recalcularSalario(arrayMusicos[i].salario);
                        }else{
                            if(arrayMusicos[i] instanceof Teclista) {
                                resultado += ((Teclista) arrayMusicos[i]).recalcularSalario(arrayMusicos[i].salario);
                            }else{
                                if(arrayMusicos[i] instanceof Percusionista) {
                                    resultado += ((Percusionista) arrayMusicos[i]).recalcularSalario(arrayMusicos[i].salario);
                                }else{
                                    if(arrayMusicos[i] instanceof CantanteGuitarrista) {
                                        resultado += ((CantanteGuitarrista) arrayMusicos[i]).recalcularSalario(arrayMusicos[i].salario);
                                    }else{
                                        resultado += ((Multinstrumental) arrayMusicos[i]).recalcularSalario(arrayMusicos[i].salario);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        return resultado;
    }

    /**
     * función que sirve para mostrar todos los musicos que hay en el array y de que tipo son
     */
    public void representarMusicosEnBanda(){
        for(int i=0;i<arrayMusicos.length;i++){
            if(arrayMusicos[i] instanceof Trompetista){
                System.out.println("Soy un trompetista");
            }else{
                if(arrayMusicos[i] instanceof Cantante){
                    System.out.println("Soy un cantante");
                }else{
                    if(arrayMusicos[i] instanceof Guitarrista){
                        System.out.println("Soy un guitarrista");
                    }else{
                        if(arrayMusicos[i] instanceof Bajista){
                            System.out.println("Soy un bajista");
                        }else{
                            if(arrayMusicos[i] instanceof Teclista) {
                                System.out.println("Soy un teclista");
                            }else{
                                if(arrayMusicos[i] instanceof Percusionista) {
                                    System.out.println("Soy un percusionista");
                                }else{
                                    if(arrayMusicos[i] instanceof CantanteGuitarrista) {
                                        System.out.println("Soy un cantnate y guitarrista");
                                    }else{
                                        System.out.println("Soy un multinstrumental");
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

File: banda/IvanRC-banda-Kotlin/src/main/kotlin/modelsBandaMusicos/Banda.kt
package modelsBandaMusicos

import factoryMusicosBanda.FactoryMusicos

class Banda(val numeroDeMusicos: Int) {

    val arrayMusicos: Array<Musico?> = Array<Musico?>(numeroDeMusicos){FactoryMusicos.getInstance()!!.createMusicoRandom()}

    /**
     * función que sirve para contar el número de trompetistas en el array
     * @return el numero de trompetistas en el array
     */
    fun contarTrompetistas(): Int{
        var contador = 0
        for(i in arrayMusicos){
            if(i is Trompetista) contador++
        }
        return contador
    }

    /**
     * función que sirve para contar el número de cantantes en el array
     * @return el numero de cantantes en el array
     */
    fun contarCantantes(): Int{
        var contador = 0
        for(i in arrayMusicos){
            if(i is Cantante) contador++
        }
        return contador
    }

    /**
     * función que sirve para contar el número de guitarristas en el array
     * @return el numero de guitarristas en el array
     */
    fun contarGuitarristas(): Int{
        var contador = 0
        for(i in arrayMusicos){
            if(i is Guitarrista) contador++
        }
        return contador
    }

    /**
     * función que sirve para contar el número de bajista en el array
     * @return el numero de bajista en el array
     */
    fun contarBajistas(): Int{
        var contador = 0
        for(i in arrayMusicos){
            if(i is Bajista) contador++
        }
        return contador
    }

    /**
     * función que sirve para contar el número de teclistas en el array
     * @return el numero de teclistas en el array
     */
    fun contarTeclistas(): Int{
        var contador = 0
        for(i in arrayMusicos){
            if(i is Teclista) contador++
        }
        return contador
    }

    /**
     * función que sirve para contar el número de percusionista en el array
     * @return el numero de percusionista en el array
     */
    fun contarPercusionistas(): Int{
        var contador = 0
        for(i in arrayMusicos){
            if(i is Percusionista) contador++
        }
        return contador
    }

    /**
     * función que sirve para contar el número de cantantesGuitarristas en el array
     * @return el numero de cantantesGuitarristas en el array
     */
    fun contarCantantesGuitarristas(): Int{
        var contador = 0
        for(i in arrayMusicos){
            if(i is CantanteGuitarrista) contador++
        }
        return contador
    }

    /**
     * función que sirve para contar el número de multinstrumentalista en el array
     * @return el numero de multinstrumentalista en el array
     */
    fun contarMultinstrumentales(): Int{
        var contador = 0
        for(i in arrayMusicos){
            if(i is Multinstrumental) contador++
        }
        return contador
    }

    /**
     * función que permite calcular el salario total
     * @return el salario total
     */
    fun hallarSalarioTotal(): Double{
        var salario = 0.0
        for(i in arrayMusicos){
            when(i){
                is Trompetista -> salario += i.recalcularSalario(i.salario)
                is Cantante -> salario += i.recalcularSalario(i.salario)
                is Guitarrista -> salario += i.recalcularSalario(i.salario)
                is Bajista -> salario += i.recalcularSalario(i.salario)
                is Teclista -> salario += i.recalcularSalario(i.salario)
                is Percusionista -> salario += i.recalcularSalario(i.salario)
                is CantanteGuitarrista -> salario += i.recalcularSalario(i.salario)
                is Multinstrumental -> salario += i.recalcularSalario(i.salario)
            }
        }
        return salario
    }

    /**
     * función que sirve para mostrar todos los musicos que hay en el array y de que tipo son
     */
    fun representarMusicosEnBanda(){
        for(i in arrayMusicos){
            when(i){
                is Trompetista -> println("Soy un Trompetista")
                is Cantante -> println("Soy un Cantante")
                is Guitarrista -> println("Soy un Guitarrista")
                is Bajista -> println("Soy un Bajista")
                is Teclista -> println("Soy un Teclista")
                is Percusionista -> println("Soy un Percusionista")
                is CantanteGuitarrista -> println("Soy un CantanteGuitarrista")
                is Multinstrumental -> println("Soy un Multinstrumental")
            }
        }
    }
}